File: docs/source/_static/examples/python/dynamical_correlation.py
"""Dynamical correlation examples."""

# --------------------------------------------------------------------------------------------
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See LICENSE.txt in the project root for license information.
# --------------------------------------------------------------------------------------------

################################################################################
# start-cell-create
from qdk_chemistry.algorithms import create

# Create a DynamicalCorrelationCalculator instance
mp2_calculator = create("dynamical_correlation_calculator")
# end-cell-create
################################################################################

################################################################################
# start-cell-configure
# Configure settings (for implementations that support them)
# mp2_calculator.settings().set("conv_tol", 1e-8)
# end-cell-configure
################################################################################

################################################################################
# start-cell-run
from pathlib import Path  # noqa: E402
from qdk_chemistry.data import Ansatz, Structure  # noqa: E402

# Load H2 structure from XYZ file
structure = Structure.from_xyz_file(Path(__file__).parent / "../data/h2.structure.xyz")

# Run initial SCF to get reference wavefunction
scf_solver = create("scf_solver")
E_hf, wfn_hf = scf_solver.run(
    structure, charge=0, spin_multiplicity=1, basis_or_guess="def2-svp"
)

# Create Hamiltonian from orbitals
hamiltonian_constructor = create("hamiltonian_constructor")
hamiltonian = hamiltonian_constructor.run(wfn_hf.get_orbitals())

# Create ansatz combining wavefunction and Hamiltonian
ansatz = Ansatz(hamiltonian, wfn_hf)

# Run the correlation calculation
mp2_total_energy, final_wavefunction = mp2_calculator.run(ansatz)

# Extract correlation energy
mp2_corr_energy = mp2_total_energy - E_hf
print(f"MP2 Correlation Energy: {mp2_corr_energy:.10f} Hartree")
print(f"MP2 Total Energy: {mp2_total_energy:.10f} Hartree")
# end-cell-run
################################################################################

################################################################################
# start-cell-list-implementations
from qdk_chemistry.algorithms import registry  # noqa: E402

print(registry.available("dynamical_correlation_calculator"))
# ['pyscf_coupled_cluster', 'qdk_mp2_calculator']
# end-cell-list-implementations
################################################################################

File: docs/source/_static/examples/cpp/dynamical_correlation.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE.txt in the project root for
// license information.

// Dynamical correlation examples.
// -----------------------------------------------------------------------------
// start-cell-create
#include <qdk/chemistry.hpp>

using namespace qdk::chemistry::algorithms;
using namespace qdk::chemistry::data;

// Create a DynamicalCorrelationCalculator instance
auto mp2_calculator =
    DynamicalCorrelationCalculatorFactory::create("qdk_mp2_calculator");
// end-cell-create
// -----------------------------------------------------------------------------

// -----------------------------------------------------------------------------
// start-cell-configure
// Configure settings (for implementations that support them)
// mp2_calculator->settings().set("conv_tol", 1e-8);
// end-cell-configure
// -----------------------------------------------------------------------------

// -----------------------------------------------------------------------------
// start-cell-run
// Load H2 structure from XYZ file
auto structure = Structure::from_xyz_file("../data/h2.structure.xyz");

// Run initial SCF to get reference wavefunction
auto scf_solver = ScfSolverFactory::create();
auto [E_HF, wfn_HF] = scf_solver->run(structure, 0, 1);

// Create Hamiltonian from orbitals
auto hamiltonian_constructor = HamiltonianConstructorFactory::create();
auto hamiltonian = hamiltonian_constructor->run(wfn_HF->get_orbitals());

// Create ansatz combining wavefunction and Hamiltonian
auto ansatz = std::make_shared<Ansatz>(*hamiltonian, *wfn_HF);

// Run the correlation calculation
auto [mp2_total_energy, final_wavefunction] = mp2_calculator->run(ansatz);

// Extract correlation energy
double mp2_corr_energy = mp2_total_energy - E_HF;
std::cout << "MP2 Correlation Energy: " << mp2_corr_energy << " Hartree\n";
std::cout << "MP2 Total Energy: " << mp2_total_energy << " Hartree\n";
// end-cell-run
// -----------------------------------------------------------------------------

// -----------------------------------------------------------------------------
// start-cell-list-implementations
auto names = DynamicalCorrelationCalculatorFactory::available();
for (const auto& name : names) {
  std::cout << name << std::endl;
}
// end-cell-list-implementations
// -----------------------------------------------------------------------------
